activerecord/lib/active_record/autosave_association.rb

```diff
@@ -155,24 +155,44 @@ def self.valid_options
 
     module ClassMethods # :nodoc:
       private
-        def define_non_cyclic_method(name, &block)
-          return if instance_methods(false).include?(name)
-          define_method(name) do |*args|
-            result = true; @_already_called ||= {}
-            # Loop prevention for validation of associations
-            unless @_already_called[name]
-              begin
-                @_already_called[name] = true
-                result = instance_eval(&block)
-              ensure
-                @_already_called[name] = false
+        if Module.method(:method_defined?).arity == 1 # MRI 2.5 and older
```

Comment:
This is a large diff and a lot of duplication, which will be dropped again next time Rails bumps its minimum Ruby version. How about conditionally defining a small private helper method instead:

```ruby
if Module.method(:method_defined?).arity == 1 # MRI 2.5
  def non_cyclic_method_already_defined?(name)
    instance_methods(false).include?(name)
  end
else
  def non_cyclic_method_already_defined?(name)
    method_defined?(name, false)
  end
end
```

Reply:
I understand the concern, I kinda had the same reaction when writing it. I'm not too fan about the private helper method though. Often time in such case people use refinements to backport Ruby features.

I'll push a refinement version.